File: src/batsh_ast.ml
open Core_kernel.Std

type identifier = string

and identifiers = identifier list

and leftvalue =
  | Identifier of identifier
  | ListAccess of (leftvalue * expression)

and expression =
  | Bool of bool
  | Float of float
  | Int of int
  | List of expressions
  | String of string
  | Leftvalue of leftvalue
  | ArithUnary of (string * expression)
  | ArithBinary of (string * expression * expression)
  | Concat of (expression * expression)
  | StrCompare of (string * expression * expression)
  | Call of (identifier * expressions)

and expressions = expression list

and statement =
  | Comment of string
  | Block of statements
  | Expression of expression
  | Assignment of (leftvalue * expression)
  | If of (expression * statement)
  | IfElse of (expression * statement * statement)
  | While of (expression * statement)
  | Global of identifier
  | Return of expression option
  | Empty

and statements = statement list

and toplevel =
  | Statement of statement
  | Function of (identifier * identifiers * statements)

and t = toplevel list
with sexp_of

File: src/BatshAst.hs
module BatshAst where

data Literal = Bool Bool
  | Int Int
  | Float Float
  | String String
  | List [Expression]
  deriving (Eq,Show)

type Identifier = String

data UnaryOperator = Not | Negate
  deriving (Eq,Show)

data BinaryOperator = Plus | Minus | Multiply | Divide | Modulo | Concat
  | Equal | NotEqual | ArithEqual | ArithNotEqual | Greater | Less
  | GreaterEqual | LessEqual | And | Or
  deriving (Eq,Show)

data LeftValue = Identifier Identifier
  | ListAccess (LeftValue, Expression)
  deriving (Eq,Show)

data Expression = LeftValue LeftValue
  | Literal Literal
  | Unary (UnaryOperator, Expression)
  | Binary (BinaryOperator, Expression, Expression)
  | Assign (LeftValue, Expression)
  | Call (Identifier, [Parameter])
  deriving (Eq,Show)

type Parameter = Expression

data Statement = Comment String
  | Block [Statement]
  | Expression Expression
  | If (Expression, Statement)
  | IfElse (Expression, Statement, Statement)
  | While (Expression, Statement)
  | Global Identifier
  | Return (Maybe Expression)
  deriving (Eq,Show)

data TopLevel = Statement Statement
  | Function (Identifier, [Identifier], [Statement])
  deriving (Eq,Show)

type Program = [TopLevel]
